### ISSUE TITLE
Setting model/layout with setxkbmap causes incorrect keymapping

### ISSUE THREAD

**Issue by rcabell** [XQuartz/XQuartz#148]
Using `setxkbmap` to use a different model / layout causes garbage characters to appear afterward.

By default, this is what `setxkbmap -query` returns:

> rules:      base
model:      empty
layout:     empty

If I try to set it to a US layout with Apple keyboard like so:

> setxkbmap -model apple -layout us

Then backspace becomes `,`, return becomes `j`, [d, f, g] becomes [`1`, `2`, `4`], etc. No way to fix it other than to restart Xquartz.

This is a problem because some X11 client applications make a call to xkb internally to set a keymap and then are unusable afterwards.

**Comment by jeremyhu** [comment]
Yep, you should let XQuartz manage its keymap.  You can probably fix it back by switching the OSX keymap, typing something in X11, then switching the OSX keymap back.

**Comment by rcabell** [comment]
> Yep, you should let XQuartz manage its keymap. You can probably fix it back by switching the OSX keymap, typing something in X11, then switching the OSX keymap back.

Unfortunately, this didn't work. I tried it with "Follow system keyboard layout" both on and off.  I also tried using `xkbcomp` and `setxkbmap -print` to save a working configuration and then set it back with `xbcomp` but that didn't work either. 

I've haven't (yet) found any work-around other than to restart Xquartz but that obviously doesn't allow applications that set the keymap to be usable...

**Comment by nbriggs** [comment]
While following the thread for the issue I reported in #246, I found myself in the xkb code in the server... where I *think* that the mapping gets set up in XkbInitKeyTypes(), but I'm not in a position to recompile XQuartz with debugging on, and I have no idea how that code is supposed to work with regard to switching the OS keymap...

**Comment by jeremyhu** [comment]
AIUI, the problem is that `setxkbmap -query` is only useful in cases where the keymap is something static on disk.  In our case, XQuartz generates the mapping dynamically by querying the macOS system keymap.

**Comment by nbriggs** [comment]
I'm not sure that I agree with that.  Sure, `setxkbmap ...` when it's trying to *set* a map is not so useful when XQuartz is generating the map on the fly -- but doing `setxkbmap -query`, or calling `XGetKeyboardMapping()` (which is likely what `setxkbmap ...` is doing), should get the same map that XQuartz would generate on the fly -- and it does not appear to do that.   I don't know if the hooks are already built in for XQuartz to trivially do the right thing (probably not), but the fact that a program using  `XGetKeyboardMapping()` gets incorrect data back is problematic.

**Comment by masinter** [comment]
should the XGetKeybaodMapping problem be a separate issue?

**Comment by nbriggs** [comment]
@masinter see #246

**Comment by 3405691582** [comment]
This problem is highlighted if you try and round-trip convert the xkb keymap obtained from the core keyboard device and back again; this works fine on other X servers. I suspect part of the problem is `setxkbmap -print` reports  keycodes as `empty+aliases(qwerty)`.

Empirically, the keycode for `a` is `8` (English, QWERTY, pc keyboard). If I get the corresponding keysym with `xkb_state_key_get_one_sym` from the `xkb_state` derived from the keymap derived from `xkb_x11_keymap_new_from_device`, then it works as expected,. If I subsequently serialize the keymap with `xkb_keymap_get_as_string` and then deserialize with `xkb_keymap_new_from_string`, then I expect the keymaps to be identical, and resolving the keycode `8` should give me the keysym `a` again, but that's not the case with XQuartz.

Workaround is to use RMLVO with rules as `base` and model as `macintosh_old`. 

Here is some sample code that highlights the problem. 

```
#include <stdio.h>
#include <stdlib.h>
#include <xcb/xcb.h>
#include <xkbcommon/xkbcommon-x11.h>
#include <xkbcommon/xkbcommon.h>

int main(int argc, char *argv[]) {
  struct xkb_context *ctx = xkb_context_new(XKB_CONTEXT_NO_FLAGS);
  if (!ctx) {
    printf("xkb_context_new\n");
    abort();
  }

  xcb_connection_t *c = xcb_connect(NULL, NULL);
  if (xcb_connection_has_error(c)) {
    printf("xcb_connect\n");
    abort();
  }

  int r = xkb_x11_setup_xkb_extension(c, XKB_X11_MIN_MAJOR_XKB_VERSION,
                                      XKB_X11_MIN_MINOR_XKB_VERSION, 0, NULL,
                                      NULL, NULL, NULL);
  if (r == 0) {
    printf("xkb_x11_setup_xkb_extension\n");
    abort();
  }

  int32_t device_id = xkb_x11_get_core_keyboard_device_id(c);
  if (device_id == -1) {
    printf("xkb_x11_get_core_keyboard_device_id\n");
    abort();
  }

  struct xkb_keymap *keymap = xkb_x11_keymap_new_from_device(
      ctx, c, device_id, XKB_KEYMAP_COMPILE_NO_FLAGS);
  if (!keymap) {
    printf("xkb_x11_keymap_new_from_device\n");
    abort();
  }

  struct xkb_keymap *k;
  struct xkb_state *state;
  xkb_keycode_t keycode = 8;
  xkb_keysym_t keysym;
  char keysym_name[64];

  k = keymap;
  state = xkb_state_new(k);
  if (!state) {
    printf("xkb_state_new\n");
    abort();
  }

  keysym = xkb_state_key_get_one_sym(state, keycode);
  xkb_keysym_get_name(keysym, keysym_name, sizeof(keysym_name));
  printf("%s\n", keysym_name);

  char *string = xkb_keymap_get_as_string(keymap, XKB_KEYMAP_FORMAT_TEXT_V1);
  k = xkb_keymap_new_from_string(ctx, string, XKB_KEYMAP_FORMAT_TEXT_V1,
                                 XKB_KEYMAP_COMPILE_NO_FLAGS);

  state = xkb_state_new(k);
  if (!state) {
    printf("xkb_state_new\n");
    abort();
  }

  keysym = xkb_state_key_get_one_sym(state, keycode);
  xkb_keysym_get_name(keysym, keysym_name, sizeof(keysym_name));
  printf("%s\n", keysym_name);

  return 0;
}
```
